packages/node-opentelemetry/src/spanProcessor.ts

```diff
@@ -0,0 +1,97 @@
+import { Attributes, Context, context, trace } from '@opentelemetry/api';
+import { BatchSpanProcessor, Span } from '@opentelemetry/sdk-trace-base';
+import { suppressTracing } from '@opentelemetry/core';
+
+import hdx from './debug';
+
+const HDX_CONTEXT_REFRESHER_INTERVAL = 10000;
+const HDX_CONTEXT_TRACE_ATTRIBUTES_EXPIRATION = 5 * 60 * 1000;
+
+class HyperDXContext {
+  private readonly _traceAttributes = new Map<string, Attributes>();
+  private readonly _traceMap = new Map<
+    string,
+    {
+      spans: Span[];
+      lastUpdateAt: number;
+    }
+  >();
+
+  constructor() {
+    // expires after 5 minutes
+    setInterval(() => {
+      hdx(`Running _traceMap expiration check`);
+      const now = Date.now();
+      for (const [traceId, data] of this._traceMap.entries()) {
+        if (now - data.lastUpdateAt > HDX_CONTEXT_TRACE_ATTRIBUTES_EXPIRATION) {
+          this._traceMap.delete(traceId);
+          this._traceAttributes.delete(traceId);
+          hdx(`Deleted traceId ${traceId} from _traceMap and _traceAttributes`);
+        }
+      }
+    }, HDX_CONTEXT_REFRESHER_INTERVAL);
+  }
+
+  private _getActiveSpanTraceId = (): string | undefined => {
+    const activeSpan = trace.getActiveSpan();
+    if (!activeSpan) {
+      return undefined;
+    }
+    return activeSpan.spanContext().traceId;
+  };
+
+  private _setTraceAttributesForAllSpans = (traceId: string): void => {
+    const attributes = this._traceAttributes.get(traceId);
+    if (!attributes) {
+      return;
+    }
+    const traceData = this._traceMap.get(traceId);
+    if (traceData) {
+      for (const span of traceData.spans) {
+        if (!span.ended) {
+          span.setAttributes(attributes);
+        }
+      }
+    }
+  };
+
+  addTraceSpan = (traceId: string, span: Span): void => {
+    // prevent downstream exporter calls from generating spans
+    context.with(suppressTracing(context.active()), () => {
+      hdx(`Adding traceId ${traceId} to _traceMap`);
+      const traceData = this._traceMap.get(traceId);
+      if (!traceData) {
+        this._traceMap.set(traceId, {
```

Comment:
Thinking if we should have some hard limit on the entries within the map and spans here as well to prevent runaway memory usage? (Not sure if this is a realistic concern, though it's possible the ref we store here could cause spans to not get GC'd?)

Reply:
Thought about it. I think we should. Although there is a refresh thing that would clear out the map time to time

Reply:
My concern is that a lot can accumulate in 5 minutes, though I guess we can get some more real-world data from dog fooding this one internally.

Reply:
I pushed the update with hard limits